crates/block-producer/src/tx_pool/mod.rs

```diff
@@ -0,0 +1,514 @@
+
+            if parent.status == TransactionStatus::InQueue {
+                return;
+            }
+        }
+
+        self.tx_roots.insert(tx_id);
+    }
+
+    /// Add the batch as a root if all parents have been placed into blocks.
+    fn try_root_batch(&mut self, batch_id: BatchId) {
+        let batch = self.batch_pool.get_mut(&batch_id).expect("Batch must be in pool");
+        for parent in &batch.parents.clone() {
+            let parent = self.batch_pool.get(parent).expect("Parent batch must be in pool");
+
+            if parent.status != BatchStatus::Blocked {
+                return;
+            }
+        }
+        self.batch_roots.insert(batch_id);
+    }
+}
+
+#[derive(thiserror::Error, Debug, Clone, PartialEq)]
+pub enum AddTransactionError {
+    #[error("Transaction's initial account state {expected} did not match the current account state {current}.")]
+    InvalidAccountState { current: Digest, expected: Digest },
+    #[error("Transaction input data is stale. Required data fresher than {stale_limit} but inputs are from {input_block}.")]
+    StaleInputs {
+        input_block: BlockNumber,
+        stale_limit: BlockNumber,
+    },
+}
+
+struct InflightTransaction {
+    status: TransactionStatus,
+    data: Arc<ProvenTransaction>,
+    parents: BTreeSet<TransactionId>,
+    children: BTreeSet<TransactionId>,
+}
+
+impl InflightTransaction {
+    /// Creates a new in-queue transaction with no children.
+    fn new(data: ProvenTransaction, parents: BTreeSet<TransactionId>) -> Self {
+        Self {
+            data: Arc::new(data),
+            status: TransactionStatus::InQueue,
+            parents,
+            children: Default::default(),
+        }
+    }
+
+    fn add_child(&mut self, child: TransactionId) {
+        self.children.insert(child);
+    }
+
+    fn batch_id(&self) -> Option<BatchId> {
+        match self.status {
+            TransactionStatus::Batched(id) => Some(id),
+            _ => None,
+        }
+    }
+
+    fn remove_parent(&mut self, parent: &TransactionId) {
+        self.parents.remove(parent);
+    }
+}
+
+struct InflightBatch {
+    status: BatchStatus,
+    transactions: Vec<TransactionId>,
+    parents: BTreeSet<BatchId>,
+    children: BTreeSet<BatchId>,
+}
+
+impl InflightBatch {
+    fn new(transactions: Vec<TransactionId>, parents: BTreeSet<BatchId>) -> Self {
+        Self {
+            status: BatchStatus::Inflight,
+            transactions,
+            parents,
+            children: Default::default(),
+        }
+    }
+
+    fn add_child(&mut self, child: BatchId) {
+        self.children.insert(child);
+    }
+
+    fn remove_child(&mut self, child: &BatchId) {
+        self.children.remove(child);
+    }
+
+    fn remove_parent(&mut self, parent: &BatchId) {
+        self.parents.remove(parent);
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum TransactionStatus {
+    InQueue,
```

Comment:
We may want to track how many times a transaction made it into a batch (e.g., as `InQueue(usize)`). This way, if let's say, the transaction was batched 3 times but failed to make it into a block, we'd automatically drop it.